sycl/include/sycl/ext/intel/experimental/pipes.hpp

```diff
@@ -0,0 +1,217 @@
+    if constexpr (_control_type == type::none) {
+      _type = 0;
+    } else if constexpr (_control_type == type::exact) {
+      _type = 1;
+    } else if constexpr (_control_type == type::max) {
+      _type = 2;
+    } else { // _control_type == type::min
+      _type = 3;
+    }
+
+    __ocl_WPipeTy<_dataT> _WPipe =
+        __spirv_CreatePipeFromPipeStorage_write<_dataT>(&m_Storage);
+    _Success = !static_cast<bool>(__latency_control_nb_write_wrapper(
+        _WPipe, &_Data, _anchor_id, _target_anchor, _type, _cycle));
+#else
+    (void)_Success;
+    (void)_Data;
+    assert(!"Pipes are not supported on a host device!");
+#endif // __SYCL_DEVICE_ONLY__
+  }
+
+  // Blocking pipes
+  // Reading from pipe is lowered to SPIR-V instruction OpReadPipe via SPIR-V
+  // friendly LLVM IR.
+  template <class... _Params> static _dataT read() {
+#ifdef __SYCL_DEVICE_ONLY__
+    static constexpr auto _anchor_id =
+        _GetValue<int32_t, latency_anchor_id, _Params...>::value;
+    static constexpr auto _constraint =
+        _GetValue3<int32_t, type, int32_t, latency_constraint,
+                   _Params...>::value;
+
+    static constexpr int32_t _target_anchor = std::get<0>(_constraint);
+    static constexpr type _control_type = std::get<1>(_constraint);
+    static constexpr int32_t _cycle = std::get<2>(_constraint);
+    int32_t _type;
+    if constexpr (_control_type == type::none) {
+      _type = 0;
+    } else if constexpr (_control_type == type::exact) {
+      _type = 1;
+    } else if constexpr (_control_type == type::max) {
+      _type = 2;
+    } else { // _control_type == type::min
+      _type = 3;
+    }
+
+    __ocl_RPipeTy<_dataT> _RPipe =
+        __spirv_CreatePipeFromPipeStorage_read<_dataT>(&m_Storage);
+    _dataT TempData;
+    __latency_control_bl_read_wrapper(_RPipe, &TempData, _anchor_id,
+                                      _target_anchor, _type, _cycle);
+    return TempData;
+#else
+    assert(!"Pipes are not supported on a host device!");
+#endif // __SYCL_DEVICE_ONLY__
+  }
+
+  // Writing to pipe is lowered to SPIR-V instruction OpWritePipe via SPIR-V
+  // friendly LLVM IR.
+  template <class... _Params> static void write(const _dataT &_Data) {
+#ifdef __SYCL_DEVICE_ONLY__
+    static constexpr auto _anchor_id =
+        _GetValue<int32_t, latency_anchor_id, _Params...>::value;
+    static constexpr auto _constraint =
+        _GetValue3<int32_t, type, int32_t, latency_constraint,
+                   _Params...>::value;
+
+    static constexpr int32_t _target_anchor = std::get<0>(_constraint);
+    static constexpr type _control_type = std::get<1>(_constraint);
+    static constexpr int32_t _cycle = std::get<2>(_constraint);
+    int32_t _type;
+    if constexpr (_control_type == type::none) {
+      _type = 0;
+    } else if constexpr (_control_type == type::exact) {
+      _type = 1;
+    } else if constexpr (_control_type == type::max) {
+      _type = 2;
+    } else { // _control_type == type::min
```

Comment:
I feel like that after `else` another `constexpr` is also needed.

Reply:
I don't think C++ allows appending `constexpr` to `else` keyword.
Instead, I eliminated the else clause by assigning an initial value to `int32_t type`.